### ISSUE TITLE
Data Availability 

### ISSUE THREAD

**Issue by marbar3778** [tendermint/tendermint#4960]
## Summary

In 0.34 state-sync and block pruning were introduced. This opens a whole can of worms as now applications need a way to enforce or incentivize a subset or all the nodes to keep the entire state. 

Prior to this Tendermint made the assumption that all nodes will have all the state.

## Problem Definition

Potential for the old state of applications to disappear. 

## Proposal

One proposal would be to introduce a new node type "Archive Node" that has a key associated with it, for different incentivization schemes. These nodes would look like any other node in the network except when there is a query for an archive node in the network they would answer. This proposal would cause changes in the gossip protocol as nodes would need to be aware of the type of node they are communicating with. This would allow applications to create a module that queries the network for nodes that are marked as "Archive" and check a sort of proof of how much state they have stored. This is **very** high-level proposal. It is meant to set a starting point for discussion.

____

#### For Admin Use

- [x] Not duplicate issue
- [x] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by adlerjohn** [comment]
Block pruning only affects _history_ (i.e. historical blocks), not _state_ (i.e. balances, contract storage). All full nodes still need to keep all state. It's important not to confound the two concepts!

For example:
> Prior to this Tendermint made the assumption that all nodes will have all the state.

I'm not aware of any assumptions in Tendermint (or Bitcoin, or Ethereum) that full nodes store historical blocks in perpetuity. Even without pruning of historical blocks implemented in the reference client, nodes could have pruned historical blocks themselves without affecting consensus.

**Comment by erikgrinaker** [comment]
>Block pruning only affects history (i.e. historical blocks), not state (i.e. balances, contract storage). All full nodes still need to keep all state. It's important not to confound the two concepts!

Absolutely. Additionally, note that e.g. transaction queries need historical blocks (as far as I know), so removing them prevents transaction histories and auditability. Also, the SDK recently introduced state pruning, which removes historical versions of state data. Just some additional aspects to keep in mind.

**Comment by musalbas** [comment]
I believe that the end goal should be to create a) a peer-to-peer distributed state storage network and b) a distributed block storage network; similar to BitTorrent or IPFS. Nodes should be able to store a small part of the state or block history, and other nodes should be able to efficiently figure out which nodes they can query to get the state or block data they require, using a system such as DHT. Ethereum developers have some discussion about this: https://ethresear.ch/t/the-data-availability-problem-under-stateless-ethereum/6973/ (note that this is to be renamed "data retrievability problem")
